src/txn-wal/src/txns.rs

```diff
@@ -543,20 +576,52 @@ where
             let _ = self.txns_cache.update_gt(ts).await;
             self.txns_cache.update_gauges(&self.metrics);
 
-            let mut unapplied_by_data = BTreeMap::<_, Vec<_>>::new();
-            for (data_id, unapplied, unapplied_ts) in self.txns_cache.unapplied() {
+            let mut unapplied_by_data =
+                BTreeMap::<_, (Arc<K::Schema>, Arc<V::Schema>, Vec<_>)>::new();
+            for (x, unapplied, unapplied_ts) in self.txns_cache.unapplied() {
                 if ts < unapplied_ts {
                     break;
                 }
-                unapplied_by_data
-                    .entry(*data_id)
-                    .or_default()
-                    .push((unapplied, unapplied_ts));
+                let (k, v, u) = unapplied_by_data.entry(x.data_id).or_insert_with(|| {
+                    (
+                        Arc::clone(&x.key_schema),
+                        Arc::clone(&x.val_schema),
+                        Vec::new(),
+                    )
+                });
+                // It's possible for a single apply_le to get work for both the
+                // old format (in which case we get a placeholder schema) and
+                // the new format (in which case we get a real schema). Use the
+                // placeholder schema if it's all we have, but if anything for
+                // this data shard has a real schema, use that instead.
+                if *k == self.placeholder_key_schema {
+                    *k = Arc::clone(&x.key_schema);
+                }
+                if *v == self.placeholder_val_schema {
+                    *v = Arc::clone(&x.val_schema);
+                }
+                // TODO: Support schema changes within txn-wal. In the meantime,
+                // assert that we got zero or one real schemas.
+                if x.key_schema != self.placeholder_key_schema {
+                    assert_eq!(*k, x.key_schema);
+                }
+                if x.val_schema != self.placeholder_val_schema {
+                    assert_eq!(*v, x.val_schema);
+                }
+                if x.key_schema == self.placeholder_key_schema
+                    || x.val_schema == self.placeholder_val_schema
```

Comment:
Oh, `UnitSchema` (used for val in prod) compares as equal to itself so this doesn't work. Could fix by only sniffing key_schema here and adding a comment or by doing something invasive to UnitSchema